### ISSUE TITLE
RuntimeError in /app/lib/meetup/api.rb:35

### ISSUE THREAD

**Issue by isometimescode** [WomenWhoCode/meet-maynard#32]
## Error in Meet-Maynard

**RuntimeError** in **/app/lib/meetup/api.rb:35**
Error parsing Meetup response: Valid Meetup Pro admin access is required. Contact Pro Support.

[View on Bugsnag](https://app.bugsnag.com/women-who-code/meet-maynard/errors/5956ed31e79ae3001891272a?event_id=5956ed31e79ae30018912729&i=gh&m=ci)

## Stacktrace

    /app/lib/meetup/api.rb:35 - rescue in get_response

[View full stacktrace](https://app.bugsnag.com/women-who-code/meet-maynard/errors/5956ed31e79ae3001891272a?event_id=5956ed31e79ae30018912729&i=gh&m=ci)

*Created automatically via Bugsnag*

**Comment by zassmin** [comment]
fixing this now 🙈

**Comment by zassmin** [comment]
fixed

**Comment by isometimescode** [comment]
An error linked to this issue has been reopened in Bugsnag
[**RuntimeError** in **/app/lib/meetup/api.rb:35**](https://app.bugsnag.com/women-who-code/meet-maynard/errors/5956ed31e79ae3001891272a?event_id=5956f041e6123e001853e482&i=gh&m=ra)

**Comment by zassmin** [comment]
now we are getting a throttle error

```
Error parsing Meetup response: Credentials have been throttled more than the allowed times in one hour
```

I thought we were handling throttling? 🤔

**Comment by isometimescode** [comment]
@zassmin we are handling throttling, but I'm not sure why it happened here. Let me take a look.

**Comment by isometimescode** [comment]
@zassmin I see the problem now. I will make a separate issue and work on a fix.

**Comment by isometimescode** [comment]
closing for now; please note that Bugsnag groups all Meetup API errors in one issue right now because its based on error class + filename + line number.

**Comment by isometimescode** [comment]
An error linked to this issue has been reopened in Bugsnag
[**RuntimeError** in **/app/lib/meetup/api.rb:35**](https://app.bugsnag.com/women-who-code/meet-maynard/errors/5956ed31e79ae3001891272a?event_id=595ed7720807e10018895a35&i=gh&m=ra)

**Comment by littleforest** [comment]
One of the issues is that our rate limit reset time is sometimes quite high, so I'm wondering if another process is starting in the meantime. For example, for this last error, our reset time was 3220 seconds = 53.6 mins, which is quite long.

Would it be worth it to set up Redis to store the reset times in, and then every process can read from the same values?

**Comment by zassmin** [comment]
side note: if we don't leverage water mark, maybe we can consider getting rid of it?

**Comment by littleforest** [comment]
We [are using it](https://github.com/WomenWhoCode/meet-maynard/blob/staging/lib/meetup/api.rb#L105-L107) to store the etag and check for the 304.

**Comment by isometimescode** [comment]
> Would it be worth it to set up Redis to store the reset times in, and then every process can read from the same values?
That or the database, but storing the data somewhere sounds like a nice simple solution in any case.

**Comment by zassmin** [comment]
> We are using it to store the etag and check for the 304.

Do we reference it for anything?

**Comment by littleforest** [comment]
@isometimescode this is a case where I think the fast look up of Redis would be superior over the db, unless we feel like the performance hit isn't worth the setup time/effort.

**Comment by littleforest** [comment]
@zassmin feeling confused about your question.  Are you just asking if we ever actually access the watermarks table.  For example, when the pro-groups task runs, it is simply forming the url as:

```
api.meetup.com/pro/womenwhocode/groups
```

It is definitely checking the database to get the etag for that query so we can first see if there is any new data to retrieve.

**Comment by littleforest** [comment]
@isometimescode also wondering if we would be more likely to run into race conditions with accessing the db vs redis?

**Comment by isometimescode** [comment]
@littleforest db or redis, either way a race condition can still occur since the amount of time between read and write of the rate limit data is dependent on the response from an external source (Meetup). If we wanted to be paranoid we'd need to implement locking to prevent a race condition. 

1. Read last ratelimit value in db|redis
2. Make request to Meetup
3. ... wait for Meetup response ... (indeterminate amount of time, since this is an external API call over the network)
4. Update ratelimit value in db|redis from Meetup response headers

**Comment by zassmin** [comment]
Per our dev sync...

Short term fix: 
* drop the db data
* schedule the length of time in which the task runs

Long term fix: 
* queuing with sidekick instead of rake tasks
* storing ratelimit in db with db lock

**Comment by zassmin** [comment]
turning this into a P1 so we can perform https://github.com/WomenWhoCode/website/issues/2197 on clean data

**Comment by isometimescode** [comment]
@zassmin just to double check, the fix here is to delete all the *production* data for the meet-maynard app, and make sure our scheduler tasks run with more time between each run? Its been so long since we talked about it, I kind of forgot what we decided 🐒

**Comment by zassmin** [comment]
that's what I recall

**Comment by isometimescode** [comment]
@WomenWhoCode/meet-maynard-reviewers I have reset the DB data and scheduled the tasks with a few hours gap between them. I also double checked that they don't overlap with other Meetup tasks from the website.

**Comment by isometimescode** [comment]
An error linked to this issue has been reopened in Bugsnag
[**RuntimeError** in **/app/lib/meetup/api.rb:35**](https://app.bugsnag.com/women-who-code/meet-maynard/errors/5956ed31e79ae3001891272a?event_id=5988bb22ef2ba70018d03d9d&i=gh&m=ra)

**Comment by zassmin** [comment]
current error: 
<img width="945" alt="image" src="https://user-images.githubusercontent.com/1654151/29042533-18674c40-7b6c-11e7-95d8-3c659fccdf3e.png">

**Comment by isometimescode** [comment]
Re-ran `rake data_import:pro_group` and there are 82 groups loaded in the database now after the task ran successfully. Next task scheduled to run will be 3:30pm PDT for `rake data_import:events`, unless you want me to start it now just to check it.

**Comment by zassmin** [comment]
@isometimescode nice! This error triggered as well... https://github.com/WomenWhoCode/meet-maynard/issues/38

do the two errors that trigger impact the next task? Are they blockers? If not let's run the events task.

**Comment by isometimescode** [comment]
@zassmin they are not blockers. #38 is just a bad programming bug. I will leave details on the issue for what the problem/fix is. I will run the events task now as well.

**Comment by isometimescode** [comment]
FYI, all the tasks seem to have run successfully *except* we did see more throttling errors last night. This is because the RSVP task ran for 6.5 hours (its the initial run). This made it cross-over the scheduled meetup tasks from the website, so I think that's why this task logged throttling errors. 

I'm not sure if #41 will give us this info, but the way I look at the timing of the tasks is by searching with `heroku/scheduler.XXXX (("Starting process") OR (State changed complete))` in Papertrail like [this log](https://papertrailapp.com/systems/1002111001/events?q=heroku%2Fscheduler.8934%20((%22Starting%20process%22)%20OR%20(State%20changed%20complete))). I got the scheduler process id by looking at the "device" tag in Bugsnag:
<img width="477" alt="screen shot 2017-08-08 at 8 48 46 am" src="https://user-images.githubusercontent.com/7094373/29081080-67e34e4a-7c16-11e7-8ebc-d2c248ac08bd.png">
